board/safety/safety_ford.h

```diff
@@ -1,112 +1,217 @@
-        tx = 0;
-      }
+
+  // Safety check for LateralMotionControl steering angle
+  if (addr == MSG_LATERAL_MOTION_CONTROL) {
+    // Signal: LatCtlPath_An_Actl
+    // Steering Angle: (0.0005 * val) - 0.5 in radians
+    // We use 0.0005 rad as a unit here
+    int raw_angle_can = ((GET_BYTE(to_send, 3) << 3) | (GET_BYTE(to_send, 4) >> 5));
+    int desired_angle = raw_angle_can - 1000;
+    // Signal: LatCtl_D_Rq
+    unsigned int steer_control_type = (GET_BYTE(to_send, 4) & 0x1CU) >> 2;
+    bool steer_control_enabled = steer_control_type > 0U;
+
+    // Rate limit while steering
+    if (controls_allowed && steer_control_enabled) {
+      // Add 1 to not false trigger violation
+      float delta_angle_float;
+      delta_angle_float = interpolate(FORD_LOOKUP_ANGLE_RATE_UP, vehicle_speed);
+      int delta_angle_up = (int)(delta_angle_float) + 1;
+      delta_angle_float = interpolate(FORD_LOOKUP_ANGLE_RATE_DOWN, vehicle_speed);
+      int delta_angle_down = (int)(delta_angle_float) + 1;
+      int highest_desired_angle = desired_angle_last + ((desired_angle_last > 0) ? delta_angle_up : delta_angle_down);
+      int lowest_desired_angle = desired_angle_last - ((desired_angle_last >= 0) ? delta_angle_down : delta_angle_up);
+
+      // Check for violation
+      violation |= max_limit_check(desired_angle, highest_desired_angle, lowest_desired_angle);
     }
-  }
+    desired_angle_last = desired_angle;
 
-  // FORCE CANCEL: safety check only relevant when spamming the cancel button
-  // ensuring that set and resume aren't sent
-  if (addr == 0x83) {
-    if ((GET_BYTE(to_send, 3) & 0x30U) != 0U) {
-      tx = 0;
+    // No angle control allowed when controls are not allowed
+    if (!controls_allowed && steer_control_enabled) {
+      violation = true;
     }
+
+    // Reset to 0 if either controls is not allowed or there's a violation
+    if (!controls_allowed || violation) {
+      desired_angle_last = 0;
+    }
+  }
+
+  // Cruise button check, only allow cancel button to be sent
+  if (addr == MSG_STEERING_DATA_FD1) {
+    // Violation if any button other than cancel is pressed
+    // Signal: CcAslButtnCnclPress
+    violation |= 0U < (GET_BYTE(to_send, 0) |
+                       (GET_BYTE(to_send, 1) & 0xFEU) |
+                       GET_BYTE(to_send, 2) |
+                       GET_BYTE(to_send, 3) |
+                       GET_BYTE(to_send, 4) |
+                       GET_BYTE(to_send, 5));
+  }
+
+  if (violation) {
+    controls_allowed = false;
+    tx = 0;
   }
 
   // 1 allows the message through
   return tx;
 }
 
-// TODO: keep camera on bus 2 and make a fwd_hook
+static int ford_fwd_hook(int bus_num, CANPacket_t *to_fwd) {
+  int addr = GET_ADDR(to_fwd);
+  int bus_fwd = -1;
+
+  switch (bus_num) {
+    case FORD_MAIN: {
+      // Forward all traffic from bus 0 onward
+      bus_fwd = FORD_CAM;
+      break;
+    }
+    case FORD_CAM: {
+      // Block stock LKAS messages (if OP is engaged)
```

Comment:
I copied this comment "if OP is engaged" from other safety code. Is this handled somewhere else in the panda?

Reply:
Not getting a hit on a search for that, I don't think it would be a correct comment for stock openpilot. See below.